stl/inc/xutility

```diff
@@ -2668,9 +2744,223 @@ namespace ranges {
     // CONCEPT ranges::contiguous_range
     template <class _Rng>
     concept contiguous_range = range<_Rng> && contiguous_iterator<iterator_t<_Rng>> && requires(_Rng& __r) {
-        { data(__r) } -> same_as<add_pointer_t<range_reference_t<_Rng>>>;
+        { _RANGES data(__r) } -> same_as<add_pointer_t<range_reference_t<_Rng>>>;
     };
     // clang-format on
+
+    // CLASS ranges::_Not_quite_object
+    class _Not_quite_object {
+    public:
+        // Some overload sets in the library have the property that their constituent function templates are not visible
+        // to argument-dependent name lookup (ADL) and that they inhibit ADL when found via unqualified name lookup.
+        // This property allows these overload sets to be implemented as function objects. We derive such function
+        // objects from this type to remove some typical object-ish behaviors which helps users avoid depending on their
+        // non-specified object-ness.
+
+        struct _Construct_tag {};
+
+        _Not_quite_object() = delete;
+
+        constexpr explicit _Not_quite_object(_Construct_tag) noexcept {}
+
+        _Not_quite_object(const _Not_quite_object&) = delete;
+        _Not_quite_object& operator=(const _Not_quite_object&) = delete;
+
+        void operator&() const = delete;
+
+    protected:
+        ~_Not_quite_object() = default;
+    };
+
+    // CLASS ranges::_Advance_fn
+    class _Advance_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <input_or_output_iterator _It>
+        constexpr void operator()(_It& _Where, iter_difference_t<_It> _Off) const {
+            if constexpr (random_access_iterator<_It>) {
+                _Where += _Off;
+            } else {

```

Comment:
In `normal` controll flow I would suggest an early return here. Is this else branch about throughput / unreachable code.

`if constexpr` is a new feature so that would be quite interesting wrt to optimizations.

Reply:
We would conventionally early-return here:
```c++
if (condition) {
  do_things;
  return;
}

do_other_things;
```
were this not a `constexpr` `if` statement to reduce indentation and generally keep things simple. Since it is a `constexpr` `if` statement, `else` branches are discarded when the `if` is taken - they aren't instantiated. In some cases that means `do_other_things()` might not even compile if we "unnested" the control flow, but even if it would compile we can improve throughput a tiny bit by avoiding instantiating what we know to be dead code.